lightly/models/spark.py

```diff
@@ -0,0 +1,280 @@
+# Code adapted from https://github.com/keyu-tian/SparK/blob/main/
+
+import math
+from typing import List, Literal, Tuple, Union
+from pprint import pformat
+import sys
+
+import torch
+import torch.nn as nn
+from lightly.models.sparse.encoder import SparseResnet
+from lightly.models.modules.sparse.spark import SparseBatchNorm2d, SparseSyncBatchNorm2d, SparseConvNeXtLayerNorm
+
+def is_pow2n(n: int) -> bool:
+    return n > 0 and (n & (n - 1)) == 0
+
+
+class UNetBlock(nn.Module):
+    def __init__(self, cin, cout, bn2d):
+        """
+        a UNet block with 2x up sampling
+        """
+        super().__init__()
+        self.up_sample = nn.ConvTranspose2d(
+            cin, cin, kernel_size=4, stride=2, padding=1, bias=True
+        )
+        self.conv = nn.Sequential(
+            nn.Conv2d(cin, cin, kernel_size=3, stride=1, padding=1, bias=False),
+            bn2d(cin),
+            nn.ReLU6(inplace=True),
+            nn.Conv2d(cin, cout, kernel_size=3, stride=1, padding=1, bias=False),
+            bn2d(cout),
+        )
+
+    def forward(self, x):
+        x = self.up_sample(x)
+        return self.conv(x)
+
+
+class LightDecoder(nn.Module):
+    def __init__(self, up_sample_ratio, width=768, sync_batch_norm=True):
+        super().__init__()
+        self.width = width
+        assert is_pow2n(up_sample_ratio)
+        n = round(math.log2(up_sample_ratio))
+        channels = [self.width // 2**i for i in range(n + 1)]
+        bn2d = nn.SyncBatchNorm if sync_batch_norm else nn.BatchNorm2d
+        self.dec = nn.ModuleList(
+            [
+                UNetBlock(cin, cout, bn2d)
+                for (cin, cout) in zip(channels[:-1], channels[1:])
+            ]
+        )
+        self.proj = nn.Conv2d(channels[-1], 3, kernel_size=1, stride=1, bias=True)
+
+        self.initialize()
+
+    def forward(self, to_dec: List[torch.Tensor]):
+        x = 0
+        for i, d in enumerate(self.dec):
+            if i < len(to_dec) and to_dec[i] is not None:
+                x = x + to_dec[i]
+            x = self.dec[i](x)
+        return self.proj(x)
+
+    def extra_repr(self) -> str:
+        return f"width={self.width}"
+
+    def initialize(self):
```

Comment:
```suggestion
    def init_weights(self):
```

Make it clear that this only initializes the weights.